clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -580,8 +601,15 @@ private Map<String, ConfigProvider> instantiateConfigProviders(Map<String, Strin
 
         for (String provider : configProviders.split(",")) {
             String providerClass = providerClassProperty(provider);
-            if (indirectConfigs.containsKey(providerClass))
-                providerMap.put(provider, indirectConfigs.get(providerClass));
+            if (indirectConfigs.containsKey(providerClass)) {
+                String providerClassName = indirectConfigs.get(providerClass);
+                if (classNameFilter.test(providerClassName)) {
```

Comment:
small suggestion here instead of having 2 nested `if`s inside a `for` maybe we can something like this 
```
Optional<String> providerClassName = Optional.ofNullable(indirectConfigs.get(providerClass));
            Boolean isAllowed = providerClassName.map(name -> classNameFilter.test(name)).orElse(false);
            if (isAllowed) {
                providerMap.put(provider, providerClassName.get());
            } else {
                throw new ConfigException(providerClassName + " is not allowed. Update System property '"
                        + AUTOMATIC_CONFIG_PROVIDERS_PROPERTY + "' to allow " + providerClassName);
            }
  ```

Reply:
This wouldn't be backwards-compatible. If the indirectConfigs.get(providerClass) is null, the existing behavior is to ignore it. TBH I think specifying a config provider alias but not a class should be a ConfigException, but not this one.